lib/socket-pool.js

```diff
@@ -1,58 +1,127 @@
+const b4a = require('b4a')
+
+const LINGER_TIME = 3000
+
 module.exports = class SocketPool {
   constructor (dht) {
     this._dht = dht
+    this._sockets = new Map()
+    this._lingering = new Set() // updated by the ref
+
+    this.routes = new SocketRoutes(this)
+  }
 
-    this._sockets = new Set()
-    this._free = new Set()
+  _onmessage (ref, data, address) {
+    this._dht.onmessage(ref.socket, data, address)
   }
 
-  _onmessage (holder, data, address) {
-    this._dht.onmessage(holder.socket, data, address)
+  _add (ref) {
+    this._sockets.set(ref.socket, ref)
   }
 
-  _add (holder) {
-    this._sockets.add(holder)
+  _remove (ref) {
+    this._sockets.delete(ref.socket)
+    this._lingering.delete(ref)
   }
 
-  _remove (holder) {
-    this._sockets.delete(holder)
-    this._free.delete(holder)
+  lookup (socket) {
+    return this._sockets.get(socket) || null
   }
 
-  _release (holder) {
-    if (this._sockets.has(holder)) this._free.add(holder)
+  setReusable (socket, bool) {
+    const ref = this.lookup(socket)
+    if (ref) ref.reusable = bool
   }
 
   acquire () {
     // TODO: Enable socket reuse
-    return new SocketHolder(this)
+    return new SocketRef(this)
   }
 
   async destroy () {
     const closing = []
 
-    for (const { socket } of this._sockets) {
-      closing.push(socket.close())
+    for (const ref of this._sockets) {
+      if (ref._timeout) clearTimeout(ref._timeout)
+      closing.push(ref.socket.close())
     }
 
     await Promise.allSettled(closing)
   }
 }
 
-class SocketHolder {
+class SocketRoutes {
+  constructor (pool) {
+    this._pool = pool
+    this._routes = new Map()
+  }
+
+  add (publicKey, rawStream) {
+    if (rawStream.socket) this._onconnect(publicKey, rawStream)
+    else rawStream.on('connect', this._onconnect.bind(this, publicKey, rawStream))
+  }
+
+  get (publicKey) {
+    const id = b4a.toString(publicKey, 'hex')
+    const route = this._routes.get(id)
+    if (!route) return null
+    return route
+  }
+
+  _onconnect (publicKey, rawStream) {
+    const id = b4a.toString(publicKey, 'hex')
+    const socket = rawStream.socket
+
+    let route = this._routes.get(id)
+
+    if (!route) {
+      route = {
+        socket,
+        address: { host: rawStream.remoteHost, port: rawStream.remotePort }
+      }
+
+      this._pool.setReusable(socket, true)
+      this._routes.set(id, route)
+
+      socket.on('close', () => {
+        if (this._routes.get(id) === route) this._routes.delete(id)
+      })
+    }
+
+    rawStream.on('error', () => {
+      this._pool.setReusable(socket, false)
+      if (!route) route = this._routes.get(id)
+      if (route && route.socket === socket) this._routes.delete(id)
+    })
+  }
+}
+
+// TODO: we should just make some "user data" object on udx to allow to attach this info
```

Comment:
`stream.userData` and `socket.userData` are available in `udx-native` 1.2.0; should we use those?

Reply:
ya let's do that in a follow up pr